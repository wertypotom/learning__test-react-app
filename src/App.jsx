import { useEffect, useReducer, useState } from 'react'
import AppSearch from './components/AppSearch'
// import ax

const initialState = {
    list: [],
    error: null
}

export const dataReducer = (state, action) => {
    if (action.type === 'SET_LIST') {
        return { ...state, list: action.payload, error: null }
    }

    if (action.type === 'SET_ERROR') {
        return { ...state, error: true }
    }

    return state
}

export const App = () => {
    const [counter, setCounter] = useState(0)
    const [data, dispatch] = useReducer(dataReducer, initialState)

    // useEffect(() => {
    //     axios.get('http://hn.algolia.com/api/v1/search?query=react')
    //         .then((response) => {
    //             const data = response.data

    //             dispatch({ type: 'SET_LIST', list: data.hits })
    //         }).catch(() => {
    //             dispatch({ type: 'SET_ERROR' })
    //         })
    // }, [])

    return (
        <div>
            <h1>My counter</h1>
            <Counter counter={counter} />

            <button type='button' onClick={() => setCounter(counter + 1)}>Increment</button>
            <button type='button' onClick={() => setCounter(counter - 1)}>Decrement</button>

            <h2>Async data</h2>
            {data.error && <div className='error'>Some error happened</div>}

            {data.list.length &&
                <ul>
                    {
                        data.list.map((el) => <li key={Date.now()}>{el.title}</li>)
                    }
                </ul>
            }

            < AppSearch />
        </div>
    )
}

export const Counter = (props) => {
    return <div><p>{props.counter}</p></div>
}