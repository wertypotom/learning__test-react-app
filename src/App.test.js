import { Counter, App, dataReducer } from './App'
import { create } from 'react-test-renderer'

describe('My first test Suite', () => {
    test('Should work', () => {
        expect(true).toBe(true)
    })
})

describe('Snapshot tests', () => {
    test('Create snapshot of App', () => {
        const component = create(<App />)

        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Create snapshot of Counter', () => {
        const component = create(<Counter counter={1} />)

        expect(component.toJSON()).toMatchSnapshot()
    })
})

describe('Test dataReducers', () => {
    it('should set data corectly', () => {
        const state = {
            list: [],
            error: null
        }
        const action = {
            type: 'SET_LIST',
            payload: [1, 2, 3, 4, 5]
        }

        const newState = dataReducer(state, action)

        /* 
            {
                list:  [1, 2, 3, 4, 5],
                error: null
            }
        */

        expect(newState).toEqual({
            list: [1, 2, 3, 4, 5],
            error: null
        })
        expect(newState.list).toEqual([1, 2, 3, 4, 5])
        expect(newState.list.length).toBe(5)
        expect(newState.error).toBeNull()
    })

    it('should correcctly set error', () => {
        const state = {
            list: [],
            error: null
        }

        const action = {
            type: 'SET_ERROR'
        }

        const newState = dataReducer(state, action)

        expect(newState).toEqual({
            list: [],
            error: true
        })
        expect(newState.list.length).toBe(0)
        expect(newState.error).toBeTruthy()
    })

    it('should remove error', () => {
        const state = {
            list: [],
            error: true
        }

        const action = {
            type: 'SET_LIST',
            payload: [1, 2, 3, 4]
        }

        const newState = dataReducer(state, action)

        expect(newState).toEqual({
            list: [1, 2, 3, 4],
            error: null
        })
        expect(newState.error).toBeNull()
        expect(newState.list.length).not.toBe(0)
    })
})