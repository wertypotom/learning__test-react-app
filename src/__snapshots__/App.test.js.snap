// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests Create snapshot of App 1`] = `
<div>
  <h1>
    My counter
  </h1>
  <div>
    <p>
      0
    </p>
  </div>
  <button
    onClick={[Function]}
    type="button"
  >
    Increment
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Decrement
  </button>
  <h2>
    Async data
  </h2>
  0
  <div>
    <ul>
      <li>
        Hello world
      </li>
      <li>
        Hello world
      </li>
      <li>
        Hello world
      </li>
    </ul>
    <div>
      <label
        htmlFor="search"
      >
        Search:
      </label>
      <input
        id="search"
        onChange={[Function]}
        placeholder="Search for something"
        type="text"
        value=""
      />
    </div>
    <p>
      Searches for 
      ...
    </p>
  </div>
</div>
`;

exports[`Snapshot tests Create snapshot of Counter 1`] = `
<div>
  <p>
    1
  </p>
</div>
`;
